### ISSUE TITLE
Migrate to NumPy 2.0

### ISSUE THREAD

**Issue by tyralla** [hydpy-dev/hydpy#147]
See the [migration guide](https://numpy.org/devdocs/numpy_2_0_migration_guide.html). This may cause us some work. However, we should first check if switching to NumPy 2.0 means problems with other libraries, such as ArcPy.  Hence, I specified NumPy < 2.0 in the requirements for now (1b2390f6d4d78fb881416be4b9a12520b071ad7a).

**Comment by tyralla** [comment]
I specified Numpy < 2.0 only in `setup.py` and `requirements.txt`.  The script `prepare_hydpy_installer.py` still writes "numpy==2..." into `make_hydpy_installer.cfg`, which is why the installer creation currently fails on AppVeyor.

**Comment by tyralla** [comment]
The newest version of ArcPy works with NumPy 1.24.3. I am afraid upgrading to 2.0 would likely cause problems. Hence, it would be good if HydPy supported NumPy 1 and 2.

The only real trouble I currently see is that we would have to support two integer types, as explained [here](https://numpy.org/doc/stable/numpy_2_0_migration_guide.html#windows-default-integer).

**Comment by tyralla** [comment]
I implemented it mainly in 4ebe9b18902e884fb358c845261a925b3a9c2721.  

To gain some safety that using NumPy 1 and 2 will work as long as expected, I ordered AppVeyor and Travis-CI in 0671613b944e72c997e24ca2491cd7fe515576d0 to run the nox doctest session mostly with NumPy2 but sometimes with NumPy 1.  The build, however, relies always on NumPy 2.  As there are at least some tests that apply Cython later (e.g., the one of class `CallbackParameter`), it might be okay to combine DLLs that have been built in combination with Numpy 1 and 2.  I also see no technical reason against it, but we should carefully check this when working with ArcPy or so.

Unfortunately, the array representations became different for NumPy 1 and 2.  This is why I finally replaced them in fe162d184ef60a6c438a02f416306f9b244b573d (it would have saved us some time if we had done it earlier).  

Due to the extended name of the doctest (`doctest(numpy='1')`) session, the temporary build path of one doctest became too long for Windows.  I do not know how to extend the allowed path length on AppVeyor, so I just changed the names used in the test a little in 8db87182c5b985f66070319d181f308cdd666908.
